docs/design/interpreter/exception-handling.md

```diff
@@ -0,0 +1,18 @@
+# CoreCLR exception handling integration with the interpreter
+There are several parts of the exception handling that need to take into consideration presence of interpreter frames on the call stack. 
+* Stack walking
+* Exception clauses enumeration
+* Exception handlers invocation
+* Resuming execution after a catch handler exits
+
+The stack walking changes are described in the [stack walking](stackwalk.md) document.
+Exception clauses enumeration in CoreCLR is agnostic of the actual low level details of EH clauses information storage. It uses an interface to a JIT manager that does the real work. As described in the stack walking document, a new InterpreterJitManager will be implemented for the interpreted code.
+To enumerate the EH clauses, it will use data provided by the InterpMethod::clauses/num_clauses.
+
+Regarding the exception handlers invocation, CoreCLR uses native helpers CallCatchFunclet, CallFinallyFunclet and CallFilterFunclet. These will need to be modified to recognize interpreted frames and call the related interpreter functions. CallFinallyFunclet would call interp_run_finally and CallFilterFunclet would call the interp_run_filter. As for the CallCatchFunclet, Mono handles calling catch funclets by calling interp_set_resume_state to record the catch handler IR code address and then returning to the mono_interp_exec_method that checks for this stored state and redirects execution to the catch handler if it is set. In CoreCLR, we may want to handle it differently, in a manner close to how interp_run_finally works.
+
+Resuming execution in the parent of a catch handler after the catch handler exits depends on whether the catch handler is in an interpreted code or in compiled managed code. 
+For the compiled code case, the existing CoreCLR EH code will handle it without changes. It would just restore the context at the resume location.
+For the interpreted code though, CoreCLR will resume execution in the mono_interp_exec_method, then pop some InterpFrame instances from the local linked list (depending on how many interpreted managed frames need to be removed) and restore the interpreter SP from the last popped one and the interpreter IP will be set to the resume location IP.
```

Comment:
I am not a big fan of context restoring into C code as we currently do on mono. I think it would be simpler to always have a C++ `try/catch` when we leave interpreter. Asking to resume into interpreter would be a simple throw. Catchers would check if the resume information is for the current block of linked interpreter frames, if it is not found then it would just rethrow. This is more or less the approach described below that we would need to take for wasm anyway.

Reply:
I didn't say that we would necessarily restore context into the C code, I said we would do that for the compiled managed code case. However, from perf point of view, it might be better to restore context back into the interp_throw on other than WASM. I have found unwinding native frames to be very costy, see my change #108480 where just getting rid of two levels of native code when propagating an exception resulted in a significant perf win.